### PR TITLE
create pass-through writer and iowriter

### PR DESCRIPTION
**Pull request by deitch**
As discussed with @SteveLasker , this is the first of 2-3 PRs expanding the capabilities of the `store` package.

There are two additions here:

* `PassthroughWriter` that lets you wrap any other underlying store (e.g. file, memory) and have a processor function in the middle. Thus if I wanted to decrypt the content in transit, or uncompress, or any other action, I have a simple framework for doing so.
* `IoWriter` is very similar to `FileStore`, except instead of taking a directory and opening files, it takes an actual `io.Writer` and passes it through. I have used it to, e.g. write directly to block devices, and to write to another stream over a network. It is more fine-grained control as you can pass it any `io.Writer` you want.

Once these are in, I have two more that take advantage of the `PassthroughWriter`: `GunzipWriter` and `UntarWriter`.

### CODE REVIEW

File: pkg/content/iowriter.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"io"
+
+	"github.com/containerd/containerd/content"
+	"github.com/opencontainers/go-digest"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// Blocksize size of byte slice for each io copy
+	Blocksize = 20480
```
Comment (jdolitsky) [inline pkg/content/iowriter.go:14]: should this be configurable?

Comment (deitch) [reply pkg/content/iowriter.go:14]: Yeah, it should. I also passed all of my logging stuff through, which is a mistake. I will clean that up too.

Comment (deitch) [reply pkg/content/iowriter.go]: OK, now it is `DefaultBlocksize` and it is an option when creating. Take a look.

**Comment by deitch** [comment]
Thanks for rebasing on master @jdolitsky . Any other comments on it?

**Comment by jdolitsky** [comment]
@deitch code lgtm - can we add a few unit tests for these? Or I suppose they can be added w/ the Gunzip/Untar writers?

**Comment by deitch** [comment]
Yeah, that is a good idea. At least for the `io.Writer` one. I will find some time to do it.

**Comment by deitch** [comment]
A bit odd to do, but I think this works for the test.

**Comment by jdolitsky** [comment]
@deitch can you check failing CI?

**Comment by deitch** [comment]
@jdolitsky I ran it multiple times locally, I cannot see the failure. Could be transient. I am repushing to kick off again

**Comment by deitch** [comment]
Nope that failed too. The failure doesn't appear to have anything to do with this PR, but I am not sure.

```
--- FAIL: TestORASTestSuite (0.27s)
    --- FAIL: TestORASTestSuite/Test_0_Push (0.09s)
    --- FAIL: TestORASTestSuite/Test_1_Pull (0.06s)
    --- PASS: TestORASTestSuite/Test_2_MediaType (0.05s)
    --- PASS: TestORASTestSuite/Test_3_Conditional_Pull (0.05s)
```

Any ideas? Here is the output https://github.com/deislabs/oras/pull/182/checks?check_run_id=1210937435

**Comment by deitch** [comment]
What is wrong with the CI run @jdolitsky ?

**Comment by deitch** [comment]
@jdolitsky I rebased onto master, but also took advantage to simplify the interface, which ended up simplifying the code and blocksize stuff materially, while hewing much more closely to the classic golang style. Take a look, and let's get this in? It has been open for a good few weeks.

File: pkg/content/iowriter.go (hunk 1 above)
Comment (deitch) [reply pkg/content/iowriter.go]: And now it is gone. It uses `io.Copy` if a blocksize is not set, an `io.CopyBuffer` if a blocksize is set. Much more golang-native-like.

**Comment by deitch** [comment]
And CI is all clean. Even better.

**Comment by deitch** [comment]
Shoot. It deleted all my changes! I have to recreate them!

**Comment by deitch** [comment]
Phew. Managed to recover it all. Back at you @jdolitsky

**Comment by deitch** [comment]
_And_ CI is green.

**Review by jdolitsky — APPROVED** [review]
lgtm!

**Comment by deitch** [comment]
Excellent.

I have gunzip and untar writers ready to go based on this.

**Pull request merged**